### ISSUE TITLE
Fallback Config infinite recursion 

### ISSUE THREAD

**Issue by cescoffier** [smallrye/smallrye-fault-tolerance#439]
Let's imagine the following class (a rest client):

```
@RegisterRestClient(configKey = "quote-service")
public interface QuoteService {

    @GET
    @Path("/")
    @CircuitBreaker
    @Timeout(2000)
    @NonBlocking
    @Fallback(fallbackMethod = "getFallbackQuote")
    Uni<String> getQuote();

    default Uni<String> getFallbackQuote() {
        return Uni.createFrom().item("No coffee - no quote");
    }

}
```

When running the application using this class, I get:

```
2021-06-13 09:00:42,418 ERROR [io.qua.run.Application] (main) Failed to start application (with profile prod): java.lang.StackOverflowError
	at io.smallrye.faulttolerance.config.FallbackConfig.isAssignableFrom(FallbackConfig.java:126)
	at io.smallrye.faulttolerance.config.FallbackConfig.isAssignableFrom(FallbackConfig.java:131)
	at io.smallrye.faulttolerance.config.FallbackConfig.isAssignableFrom(FallbackConfig.java:131)
	at io.smallrye.faulttolerance.config.FallbackConfig.isAssignableFrom(FallbackConfig.java:131)
	at io.smallrye.faulttolerance.config.FallbackConfig.isAssignableFrom(FallbackConfig.java:131)
	at io.smallrye.faulttolerance.config.FallbackConfig.isAssignableFrom(FallbackConfig.java:131)
	at io.smallrye.faulttolerance.config.FallbackConfig.isAssignableFrom(FallbackConfig.java:131)
	at io.smallrye.faulttolerance.config.FallbackConfig.isAssignableFrom(FallbackConfig.java:131)
	at io.smallrye.faulttolerance.config.FallbackConfig.isAssignableFrom(FallbackConfig.java:131)
	at io.smallrye.faulttolerance.config.FallbackConfig.isAssignableFrom(FallbackConfig.java:131)
	at io.smallrye.faulttolerance.config.FallbackConfig.isAssignableFrom(FallbackConfig.java:131)
	at io.smallrye.faulttolerance.config.FallbackConfig.isAssignableFrom(FallbackConfig.java:131)
```

It works if I use a FallbackHandler.
Is it expected?

Env:
* Quarkus 2.0.0.CR3
* Java 11

**Comment by Ladicek** [comment]
Certainly not expected :-) I'll take a look.

**Comment by Ladicek** [comment]
The reproducer here work just fine. Fortunately the bug is obvious from the code pointed to by the stack trace. To actually reproduce, one needs to have this:

```java
@RegisterRestClient(baseUri = "http://localhost:8080")
public interface HelloClient {
    @GET
    @Path("/ft-hello")
    @NonBlocking
    @Fallback(fallbackMethod = "fallback")
    Uni hello();

    default Uni<String> fallback() {
        return Uni.createFrom().item("Hello fallback!");
    }
}
```

That is, the guarded method returns a raw type, while the fallback method returns a parameterized type. Interestingly, if you have the opposite (guarded method returning a parameterized type and the fallback method returning a raw type), you get a different exception (still wrong, though).

**Comment by cescoffier** [comment]
Weird, my method was returning a Uni<String>.
It was: https://github.com/cescoffier/reactive-systems-in-java/blob/master/chapter-12/api-gateway-example/api-gateway/src/main/java/org/acme/gateway/QuoteService.java

**Comment by Ladicek** [comment]
Should be fixed in the `main` branch now.

**Comment by Ladicek** [comment]
That said, if your guarded method and fallback method had the same return type, yet you were still running into this problem, it might have been a different issue.

One question: was this in native? If so, maybe we're missing a reflection registration somewhere.

**Comment by cescoffier** [comment]
No, was in dev mode. I will try to get a reproducer in a few minutes.

**Comment by Ladicek** [comment]
Thanks!

**Comment by cescoffier** [comment]
Here you go: https://github.com/cescoffier/reactive-systems-in-java/tree/ft-reproducer/chapter-12/api-gateway-example/api-gateway.

`mvn clean quarkus:dev` reproduce the infinite recursion issue I've seen.

**Comment by Ladicek** [comment]
Ah ah -- I couldn't reproduce because I used classic RESTEasy. RESTEasy Reactive does some weird things. I'll check.

**Comment by Ladicek** [comment]
Right, so what happens is: at build time, RESTEasy Reactive's RestClient creates a bean class that implements the RestClient interface. This bean is then used if you inject the RestClient interface, and it's also this bean that is intercepted by Fault Tolerance. Here's the catch: when copying RestClient methods to this class, their generic signatures aren't copied (booo!). That's why the method essentially looks like it returns `Uni`, even though it returns `Uni<String>`.

I'll take a look if that can be rectified somehow.

**Comment by Ladicek** [comment]
So I think reconstructing the generic signature from Jandex's `MethodInfo` should be possible, but the fix from #441 seems to be effective too.

**Comment by Ladicek** [comment]
This PR https://github.com/quarkusio/quarkus/pull/17984 is a proper fix.
